source/link/linker.cpp

```diff
@@ -0,0 +1,755 @@
+  }
+  ++id_bound;
+  if (id_bound > 0x3FFFFF)
+    return libspirv::DiagnosticStream(position, impl_->context->consumer,
+                                      SPV_ERROR_INVALID_ID)
+           << "The limit of IDs, 4194303, was exceeded:"
+           << " " << id_bound << " is the current ID bound.";
+
+  // Extract linking information
+  struct LinkingData {
+    size_t moduleIndex;
+    SpvId id;
+    SpvId typeId;
+    const char* name;
+    std::vector<SpvId> parametersIds;
+  };
+  std::vector<std::vector<LinkingData>> modulesImports(modules.size());
+  std::vector<std::unordered_map<std::string, LinkingData>> modulesExports(modules.size());
+  size_t moduleIndex = 0u;
+  for (auto& i : modules) {
+    // Figure out the imports and exports
+    for (auto& j : i->annotations()) {
+      if (j.opcode() == SpvOpDecorate && j.GetSingleWordInOperand(1u) == SpvDecorationLinkageAttributes) {
+        uint32_t type = j.GetSingleWordInOperand(3u);
+        SpvId id = j.GetSingleWordInOperand(0u);
+        LinkingData data;
+        data.moduleIndex = moduleIndex;
+        data.name = reinterpret_cast<const char*>(j.GetInOperand(2u).words.data());
+        data.id = id;
+        data.typeId = 0u;
+        for (const auto&k : *i) {
+          if (k.result_id() != id)
+            continue;
+          k.ForEachParam([&data](const Instruction* inst) {
+            data.parametersIds.push_back(inst->result_id());
+          });
+        }
+        if (type == SpvLinkageTypeImport)
+          modulesImports[moduleIndex].push_back(data);
+        else if (type == SpvLinkageTypeExport)
+          modulesExports[moduleIndex].emplace(data.name, data);
+      }
+      // TODO(pierremoreau): Can an SpvOpDecorateId be used for
+      //                     LinkageAttributes decorations?
+    }
+    ++moduleIndex;
+  }
+
+  //Find the import/export pairs
+  std::vector<std::pair<LinkingData, LinkingData>> linkingsToDo;
+  for (const auto& i : modulesImports) {
+    for (const auto& j : i) {
+      std::vector<LinkingData> possibleExports;
+      for (const auto& k : modulesExports) {
+        const auto& l = k.find(j.name);
+        if (l != k.end())
+          possibleExports.push_back(l->second);
+      }
+      if (possibleExports.empty())
+        return libspirv::DiagnosticStream(position, impl_->context->consumer,
+                                          SPV_ERROR_INVALID_BINARY)
+               << "No export linkage was found for \"" << j.name << "\".";
+      else if (possibleExports.size() > 1u)
+        return libspirv::DiagnosticStream(position, impl_->context->consumer,
+                                          SPV_ERROR_INVALID_BINARY)
+               << "Too many export linkages, " << possibleExports.size()
+               << ", were found for \"" << j.name << "\".";
+
+      linkingsToDo.emplace_back(j, possibleExports.front());
+    }
+  }
+
+  // Grab all types used by imports and exports
```

Comment:
Reading this far, I don't know why this is necessary.  Add a comment saying why you need to do this.
Is it to catch a user error of trying to mismatched types?

Also, need a comment saying "now we fill in the typeId member of each linkage entry. (if I read correctly up until now)